### PR TITLE
Add support for channel tabs

### LINKED ISSUE #227: Playlist in channels
I am currently working on bandcamp support [in my fork](https://github.com/fynngodau/NewPipeExtractor). Even though it was difficult to understand at first, I feel like I now know what I'm doing.

I noticed that `ChannelExtractor.getInitialPage()` must return an `InfoItemsPage` of type `<StreamInfoItem>`. This means that channels can't contain playlist items, unlike search results, where the return type is `InfoItemsPage<InfoItem>`.

For bandcamp support, this is inconvenient, as streams (tracks) don't appear on their channel (artist's profile) themselves when they are part of a playlist (an album).

That's why I'd like to ask this to be changed so that channels can also contain playlists.

### PR DESCRIPTION
**Pull request by Theta-Dev**
- [X] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [X] I have tested the API against [NewPipe](https://github.com/TeamNewPipe/NewPipe).
- [X] I agree to create a pull request for [NewPipe](https://github.com/TeamNewPipe/NewPipe) as soon as possible to make it compatible with the changed API.

YouTube is planning to update their channel page layout (as announced [here](https://www.youtube.com/watch?v=czIyqEC4V-s)) and divide videos into 3 tabs (regular videos, shorts, livestreams).

I did open pull request #944 to address the updated video tab layout, but that would still leave shorts and livestreams inaccessible. So I decided to implement channel tabs for NewPipe that allow the extraction of additional content besides the main video list.

Channel tabs may contain any `InfoItem`, which makes it possible to not only extract video tabs but also playlists and featured channels. So this also resolves, for the extractor part, TeamNewPipe/NewPipe#2414.

I added support for the following services and tabs:
- **YouTube:** Shorts, Livestreams, Playlists, Channels, About
- **Soundcloud:** Playlists, Albums
- **Peertube:** Channels, Playlists
- **Bandcamp:** Albums

Closes #227

### CODE REVIEW

**Comment by B0pol** [comment]
Are YouTube music / topic channels supported? https://github.com/TeamNewPipe/NewPipeExtractor/issues/501
We couldn't support them because of no tab support, and it's basically a channel with playlist tab only

**Comment by FireMasterK** [comment]
I rebased and fixed conflicts at https://github.com/FireMasterK/NewPipeExtractor/tree/channel-tabs

Feel free to merge it in here :)

**Review by TobiGr — COMMENTED** [review]
I briefly went through the most important parts of the code. I think those look good.

File: extractor/src/main/java/org/schabi/newpipe/extractor/linkhandler/ChannelTabs.java
Comment (Theta-Dev) [reply extractor/src/main/java/org/schabi/newpipe/extractor/linkhandler/ChannelTabs.java:12]: I used an enum in the initial version, but decided to use the existing LinkHandlers which take strings for content filtering. The search function works the same.

Comment (FireMasterK) [reply extractor/src/main/java/org/schabi/newpipe/extractor/linkhandler/ChannelTabs.java:12]: Can't you use an enum and call `.name()` on it?

Comment (Theta-Dev) [reply extractor/src/main/java/org/schabi/newpipe/extractor/linkhandler/ChannelTabs.java:12]: Yes, that is an option. But I dont see an advantage in that. We wont get the type safety of an enum and it would make the code longer.

There arent any enums used for search filters, either.

Comment (FireMasterK) [reply extractor/src/main/java/org/schabi/newpipe/extractor/linkhandler/ChannelTabs.java:12]: Well, sure, but we can check that they're valid by iterating over `Enum.values()`. It's more to keep the code than actual type safety clean I would say.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java
Comment (FireMasterK) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java:603]: Please revert these formatter changes!

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java
Comment (FireMasterK) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java:687]: This too

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java
Comment (FireMasterK) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java:698]: This too

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java
Comment (FireMasterK) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java:848]: This too

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java
Comment (FireMasterK) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java:1262]: You've changed the code here, but the formatting needs to be changed here

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java
Comment (FireMasterK) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java:1317]: This too

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java
Comment (FireMasterK) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java:1349]: This too

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java
Comment (FireMasterK) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java:1377]: This too

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java
Comment (FireMasterK) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java:1406]: This too

**Review by FireMasterK — APPROVED** [review]
Minor formatting changes that need to be addressed, otherwise this is good to merge IMO!

**Comment by Theta-Dev** [comment]
Okay, I fixed the issue with shorts continuations not loading.

Turns out that YouTube expects the visitor data to be set when getting shorts continuations. Otherwise the API simply returns an empty response.

**Comment by opusforlife2** [comment]
> Turns out that YouTube expects the visitor data to be set when getting shorts continuations. 

Oh no. More tracking?

**Comment by Theta-Dev** [comment]
Because YouTube generates a new visitorData cookie for every anonymous request, they can only associate the initial tab request with the continuations. And because no additional user data is submitted on continuation requests, there should not be any tracking potential beyond IP addresses/server logs.

File: extractor/src/main/java/org/schabi/newpipe/extractor/linkhandler/ChannelTabs.java
Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/linkhandler/ChannelTabs.java:12]: Mmmh, I agree with @FireMasterK that `Enum.values()` can be useful

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeChannelExtractor.java
Comment (Stypox) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeChannelExtractor.java:232]: Aren't you calling `getVideoTab()` twice? You are doing it both here and in `getInitialPage()`. I think you should just populate tabs in `onFetchPage()` and then use the extracted data both in `getTabs()` and `getInitialPage()`.

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/peertube/PeertubeAccountExtractorTest.java
Comment (Stypox) [inline extractor/src/test/java/org/schabi/newpipe/extractor/services/peertube/PeertubeAccountExtractorTest.java:221]: TODO detected

**Review by Stypox — CHANGES_REQUESTED** [review]
Thank you! Code looks mostly good and is structured well.

The only thing that didn't convince me is the fact that the videos tab is hardcoded as the main tab and is not actually a tab. I get that this is mostly to abide by the fact that `ChannelExtractor` extends `ListExtractor`, but it would be better if this was changed to reduce complexity and be more consistent. I think it's perfectly fine to do so in a follow-up PR.

**Comment by FireMasterK** [comment]
Something that I noticed was we can't see the full view counts on the shorts tab.

Currently, we parse them as just a regular number which is a problem.

<details>
<summary>Comparison</summary>


</details>

**Review by Stypox — APPROVED** [review]
Looks good to me, thank you! I read most of the code again and did not find issues. I only found a few minor code-style issues (for example, try to use streams more often!).

Talking about https://github.com/TeamNewPipe/NewPipeExtractor/pull/951#discussion_r1059493323: I saw you added javadocs to the function being called, which is a good choice, but could you also add a small comment on the call-site there, explaining that the params being used there is for the "Videos" tab?

I wanted to make explicit that the final decision on https://github.com/TeamNewPipe/NewPipeExtractor/pull/951#discussion_r1023332851 seems to be not to use enums but rather keep strings. This is fine for now, since it's being done all over the place in the extractor (the link handlers code design is not the best), but it's not optimal.

> The only thing that didn't convince me is the fact that the videos tab is hardcoded as the main tab and is not actually a tab. I get that this is mostly to abide by the fact that ChannelExtractor extends ListExtractor, but it would be better if this was changed to reduce complexity and be more consistent. I think it's perfectly fine to do so in a follow-up PR.

Do you have any thoughts on the above statement?

**Comment by Theta-Dev** [comment]
@Stypox I did not want to introduce a breaking change when adding channel tabs, so I kept the ChannelExtractors mostly as is and added the ChannelTabExtractors for the new content.

**Comment by AudricV** [comment]
Somme changes needs to be done in this PR, I will list them in a review which should come in the next days.

**Comment by Theta-Dev** [comment]
Okay. I think the API change makes sense. I'll fix the tests as well as the shorts upload date extraction.

**Comment by Stypox** [comment]
Oh ok, thank you. I was going to fix tests myself but if you are already on it I'll let you do that. I will add some documentation though.

Do you think there is a way to extract shorts upload date? I looked at the json of shorts but YouTube doesn't seem to provide such information.

**Comment by Stypox** [comment]
On the website Bandcamp has a single channel tab with mixed tracks and albums. Since our `ChannelTabExtractor`s extend `ListInfo<InfoItem>`, we don't need to split it in two tabs but can keep a single tab with mixed content. What do you think?

**Review by Stypox — CHANGES_REQUESTED** [review]
Some comments that need to be thought about and/or solved

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeChannelVideosTabExtractor.java
Comment (Theta-Dev) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeChannelVideosTabExtractor.java:1]: yes, it can. It was a seperate extractor with the old API since it had to extract video streams only

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java
Comment (fynngodau) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java:129]: Yes, we definetly should display albums and single tracks on the same page! Unlike with YouTube this causes no duplicates because playlists can only contain tracks that are not also displayed as a single track. (Except if an artist publishes the same track twice, which is besides the point.)

Comment (Theta-Dev) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java:129]: I would not merge the two tabs. It would not fit into our tab type scheme (we have album and track tabs, how should we call the merged one?).
There would also be issues with the playlist controls to play all tracks of a channel.

Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java]: No, don't worry, the client part is simple to fix (a `.filter(StreamInfoItem.class::isinstance)` on the returned items is enough). So I would keep one tab.

Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java]: Actually, I don't know. What is @AudricV's opinion?

Comment (fynngodau) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java]: I think if split it up, it would be more effort for a user to explore an artist's creations, because – unlike YouTube – there is no difference in how it is used between a track and an album; on YouTube you look at the Videos tab to see the latest videos and at the Playlists tab to specifically look for videos on a certain topic. Whereas on Bandcamp it doesn't matter whether your next favorite song is part of an album or not.

There's no reason we should make a distinction where the website doesn't make one.

> we have album and track tabs, how should we call the merged one?

Discography is the term we're looking for.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeAccountExtractor.java
Comment (TobiGr) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeAccountExtractor.java:166]: That looks suspicious. If the corresponding `PeertubeChannelLinkHandlerFactory` does not produce a `LinkHandler` that returns the correct ID, there might be an error somewhere else.
@Theta-Dev I run the currently implemented tests and did not run into the `else` branch. Can you give us instructions on how to get into the else branch?

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java
Comment (Theta-Dev) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java]: With Bandcamp, we are not fetching the desktop page but are using the mobile API which returns a list of discographs (both tracks and albums). The album items retuned by the API do not contain a track list, so it would not be possible (without further requests) do display all tracks of an artist in a single list.

We could merge the items together and have a list of both tracks and albums. But I dont think that would be very intuitive. Clicking "Play all" for example would only play the tracks and ignore the albums. If an artist does not have single tracks, the feature would not work at all and the user might get confused, since there are items present in the list.

Comment (fynngodau) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java]: That sounds like a matter of adapting Play All to work with playlists in my eyes, if that is the only unintuitive thing you see here. This would result in the behavior that a user expects in all regards (i.e. tracks and albums on one page like on the website, as well as being able to play all of an artist).

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeAccountExtractor.java
Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeAccountExtractor.java:166]: I just copied it over from https://github.com/TeamNewPipe/NewPipeExtractor/blob/dev/extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeAccountExtractor.java#L164-L169, which apparently was introduced by me for backwards compatibilty in https://github.com/TeamNewPipe/NewPipeExtractor/pull/309/commits/d130fd79c31338293134f8416500f135fc57a63a in 2020. If I remember correctly, the problem was that NewPipeExtractor used to provide IDs without "/accounts" before channel+accounts support was added, but now 3 years have passed, so yes, this check can be removed.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java
Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java]: I agree it would be off-topic to make the Play All button behave properly, so ok, let's keep them two separate tabs (but they should be made so that the channel page is fetched only once).

Comment (Theta-Dev) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java]: This is already the case, bandcamp channels retun up to 2 ReadyChannelTabListLinkHandlers

**Review by Stypox — APPROVED** [review]
Looks good to me. Now we just need to wait for AudricV's review and we are done. Thanks!

**Comment by Stypox** [comment]
Mmmh, you are right, but then the two urls should be ".../tracks" and ".../albums", for consistency. Can you do this change?

**Comment by Theta-Dev** [comment]
I changed the URL suffixes to `track` and `album`, as these both redirect to the main page instead of leading to a 404 error.

**Review by AudricV — CHANGES_REQUESTED** [review]
First of all, before reading the review, don't be afraid: I don't ask you to change everything by yourself, I can do a part of my requested changes if you want.
As some code changes have been made by @Stypox, they should reply for comments on their changes, so you probably don't need to answer on every comment.
Note that a lot of review comments are asking for code style or formatting changes, so changes are easier that what you can think on a first look.

Thank you very much for your effort on this feature! The general structure and code changes look fine with the current extractor code, but improvements are definitively needed in the future (and that's what #904 does partially).

We need a list of API changes which will be added in the next extractor release, especially because this PR introduces breaking changes. It should be added in the PR description for easier access.

I noticed changes that should be made globally/in multiple files (some changes are not listed here but on review comments):

- better documentation/addition of documentation on methods and classes, at least for the ones exposed in the API;
- proper/better usage of Java's Stream API (I also highlighted this in a few review comments): when working on streams of `JsonArray`s, in most cases you filter/keep only objects of `JsonObject`'s class instance then casting and working with these `JsonObject`s instead of doing the same process or parts of it multiple times (e.g. multiple casts). Here is an example:
  ```java
  myJsonArray.stream()
      .filter(JsonObject.class::isInstance)
      .map(JsonObject.class::cast)
      // Your operations here
  ```
- on several tests, you are removing test `Override` annotations, replacing them by `Test` ones. Can you explain why are you doing this?

Finally, it seems that some commits have a message different of what they are doing. Could you fix that and also, if possible, switching to a rebase on our `dev` branch instead of using a lot of merge commits and fixing merge conflicts on the fly? It would make the commits of your branch, and so this PR, a lot cleaner in my opinion. Thank you in advance!

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/SoundcloudParsingHelper.java
Comment (FireMasterK) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/SoundcloudParsingHelper.java:343]: Fyi, `.forEach()` will work the same way as `.forEachOrdered()` unless you call `.parallel()` before it.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamInfoItemExtractor.java
Comment (FireMasterK) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamInfoItemExtractor.java:1]: In such a case, I think we should add an interface / abstract class.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeChannelExtractor.java
Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeChannelExtractor.java:262]: Previously a lambda was used, which requires final variables, and then we forgot to remove the workaround

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeChannelExtractor.java
Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeChannelExtractor.java:231]: I think "addTab" is a perfectly reasonable name for a lambda that adds a tab

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamInfoItemExtractor.java
Comment (AudricV) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamInfoItemExtractor.java:1]: Why would this be needed?

**Comment by Theta-Dev** [comment]
> on several tests, you are removing test Override annotations, replacing them by Test ones. Can you explain why are you doing this?

Tests do not seem to run when they are not annotated with `@Test`. I could add both annotations, though.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamInfoItemExtractor.java
Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamInfoItemExtractor.java:1]: I don't think so

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeChannelTabExtractor.java
Comment (Theta-Dev) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeChannelTabExtractor.java:156]: I assume you mean applying the channel name extraction fix of #1050 to the ChannelTabExtractor

Comment (AudricV) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeChannelTabExtractor.java:156]: I meant using a better name like `metadataName` instead of `mdName`, but you will have to apply your changes here indeed.

**Review by FireMasterK — COMMENTED** [review]
One thing I don't like about the YouTube implementation is that we fetch the videos tab, but we don't extract the `StreamInfoItem` contents from it, just the tabs.

This requires 2 (potentially the same) requests now to get the video tab information, which makes things a lot slower for Piped :/

Could we somehow have something like a default tab extracted if the service needs to extract a tab in order to get the available tabs?

**Comment by AudricV** [comment]
Please do not push any commits to the branch of this pull request, I am working on rebasing this branch on top of the `dev` branch and applying the remaining requested changes. Thanks!

**Comment by AudricV** [comment]
Closing in favor of #1082. Thank you very much for your work which you can find in this new, improved and updated PR of this one!

**Pull request closed without merging**
